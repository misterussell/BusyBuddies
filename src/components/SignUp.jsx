import React, { useState } from 'react';
import PropTypes from 'prop-types';

function SignUp(props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState(null);
  const [confirm, setConfirm] = useState(null);
  const [confirmationCode, setConfirmationCode] = useState('');


  const errorDiv = error
    ? (
      <div className="sign-up-error">
        <p>{error}</p>
      </div>
    )
    : null;

  const confirmDiv = confirm
    ? (
      <div className="confirmation">
        <p>Please check your email for your confirmation code.</p>
        <p>Code:</p>
        <input
          value={confirmationCode}
          onChange={handleConfirmationCode}
          placeholder="code"
          label="code"
        />
        <button onClick={handleConfirm}>confirm</button>
      </div>
    )
    : null;

  return (
    <div className="sign-up">
      <form>
        <p>email:</p>
        <input
          value={email}
          onChange={handleEmailChange}
          placeholder="email"
          label="email"
        />
        <p>password:</p>
        <input
          value={password}
          onChange={handlePasswordChange}
          placeholder="password"
          label="password"
        />
        <button
          onClick={handleSubmit}
          label="submit"
          >sign up</button>
      </form>
      {
        errorDiv
      }
      {
        confirmDiv
      }
    </div>
  );

  function handleEmailChange(e) {
    setEmail(e.target.value);
  }

  function handlePasswordChange(e) {
    setPassword(e.target.value);
  }

  function handleSubmit(e) {
    e.preventDefault();
    // props.user
    props.user.signUp('max@misterussell.com', 'NewUser1!').then(response => {
      console.log('no error')
      setConfirm(true);
      setError(null);
    }).catch(error => {
      setError(error.message);
    });
  }

  // handleConfirm = (e) => {
  //   e.preventDefault();
  //   const email = this.state.email.trim();
  //   const code = this.state.confirmationCode.trim();
  //   // confirm account
  // }

}

SignUp.propTypes = {
  store: PropTypes.object,
};

export default SignUp;
