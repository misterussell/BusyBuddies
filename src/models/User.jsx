import {
  CognitoUser,
  CognitoUserPool,
  CognitoUserAttribute,
  AuthenticationDetails,
} from 'amazon-cognito-identity-js';

import cognitoConfig from '../CognitoConfig';

export default class User {
  constructor() {
    this.isAuthenticated = false;
    this.userPool = new CognitoUserPool({
      UserPoolId: cognitoConfig.UserPoolId,
      ClientId: cognitoConfig.ClientId
    });
  }

  isAuthenticated = null;

  setAuthenticated() {
    this.isAuthenticated = true;
  }

  signUp(email, password) {
    const attributeList = [
      new CognitoUserAttribute({
        Name: 'email',
        value: email,
      })
    ];

    return new Promise((resolve, reject) => {
      if (email && password) {
        this.userPool.signUp(email, password, attributeList, null, (err, result) => {
          if (err) {
            reject(err);
          } else resolve(result.user);
        });
      } else {
        reject({
          error: 'User not signed up. Email and password required.'
        });
      };
    });
  }

  confirm(email, code) {
    const user = {
      Username: email,
      Pool: this.userPool,
    };
    const cognitoUser = new CognitoUser(user);
    return new Promise((resolve, reject) => {
      if (email && code) {
        cognitoUser.confirmRegistration(code, true, (err, result) => {
          if (err) {
            reject(err.message || JSON.stringify(err));
          } else resolve(result);
        });
      } else {
        reject({
          error: 'User not confirmed. Email and code required.'
        });
      };
    });
  }

  signIn(email, password) {
    const authData = {
      Username: email,
      Password: password,
    };
    const cognitoAuthDetails = new AuthenticationDetails(authData);
    const user = {
      Username: email,
      Pool: this.userPool,
    };
    const cognitoUser = new CognitoUser(user);
    return new Promise((resolve, reject) => {
      if (email && password) {
        cognitoUser.authenticateUser(cognitoAuthDetails, {
          onSuccess: function(result) {
            resolve(result.getAccessToken().getJwtToken());
          },
          onFailure: function(error) {
            reject(error);
          },
        });
      } else {
        reject({
          error: 'User not signed in. Email and password required.'
        });
      }
    });
  }

  signOut() {
    const currentUser = this.userPool.getCurrentUser();
    const signOut = (user) => user ? currentUser.signOut() : null;
    return signOut(currentUser);
  }

  getSession() {
    const user = this.userPool.getCurrentUser();
    return new Promise((resolve, reject) => {
      if (user) {
        user.getSession((error, result) => {
          if (error) reject(error);
          if (result) {
            resolve(result.getAccessToken().getJwtToken());
            this.setAuthenticated();
          };
        });
      } else {
        reject('No user session found.');
      };
    });
  }

}
