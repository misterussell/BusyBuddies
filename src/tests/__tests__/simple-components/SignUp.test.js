import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import SignUp from '../../../components/SignUp';
import User from '../../../models/User';

const user = new User();

beforeEach(() => {
  const div = document.createElement('div');
  ReactDOM.render(<SignUp user={ user }/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('the SignUp component', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SignUp user={ user }/>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders a SignUp component correctly', () => {
    const SignUpComponent = renderer.create(<SignUp user={ user }/>).toJSON();
    expect(SignUpComponent).toMatchSnapshot();
  });

  it('should have access to the user instance from store', () => {
    // do I need to test this with propTypes?
  });

  it('should render two inputs and a button initially', () => {
    // while awaitConfirm is null the password input is shown
    // when the signUp form is submit the the password is hidden
    // if the user chooses to confirm later the password input is hidden
    // inputs with classNames: email-input, password-input
    // button className: submit-button
  });

  it('should render two inputs and a button when confirming account', () => {
    // when awaitConfirm is set to true the password input is unmounted
    // and the confirm input is mounted
  });

  it('should render an error for each invalid from field', () => {
    // when an form validation error is stored to state
    // an error div should be rendered
  });
});
