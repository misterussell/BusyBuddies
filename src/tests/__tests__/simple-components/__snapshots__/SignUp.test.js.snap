// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the SignUp component renders a SignUp component correctly 1`] = `
<div
  className="sign-up"
>
  <form>
    <p>
      email:
    </p>
    <input
      className="email-input"
      label="email"
      onChange={[Function]}
      placeholder="email"
      value=""
    />
    <div
      className="sign-up password"
    >
      <p>
        password:
      </p>
      <input
        className="password-input"
        label="password"
        onChange={[Function]}
        placeholder="password"
        value=""
      />
    </div>
    <button
      className="submit-button"
      label="submit"
      onClick={[Function]}
    >
      sign up
    </button>
    <button
      onClick={[Function]}
    >
      I already have a code
    </button>
  </form>
</div>
`;
